diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3a48eaf4e29afa60dbfddfd91f8910ca40844da5..ca3368184ccff83b29f3e996388d2883abdd644b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,12 +40,17 @@ endif()
 set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard whose features are requested to build this target.")
 set(CMAKE_C_STANDARD   11 CACHE STRING "The C standard whose features are requested to build this target.")
 
+if(NOT DEFINED GLIBCXX_USE_CXX11_ABI)
+  set(GLIBCXX_USE_CXX11_ABI 0)
+endif()
+
 # ---[ Utils
 include(cmake/public/utils.cmake)
 
 if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   include(cmake/CheckAbi.cmake)
   string(APPEND CMAKE_CXX_FLAGS " -D_GLIBCXX_USE_CXX11_ABI=${GLIBCXX_USE_CXX11_ABI}")
+  message(STATUS "CMAKE_CXX_FLAGS: ${CMAKE_CXX_FLAGS}")
   string(APPEND CMAKE_CUDA_FLAGS " -D_GLIBCXX_USE_CXX11_ABI=${GLIBCXX_USE_CXX11_ABI}")
   if(${GLIBCXX_USE_CXX11_ABI} EQUAL 1)
     set(CXX_STANDARD_REQUIRED ON)
