diff --git a/torch/nn/parallel/data_parallel.py b/torch/nn/parallel/data_parallel.py
index 44e571e728..48a29ed55a 100644
--- a/torch/nn/parallel/data_parallel.py
+++ b/torch/nn/parallel/data_parallel.py
@@ -24,6 +24,11 @@ def _check_balance(device_ids):
     device_ids = [_get_device_index(x, True) for x in device_ids]
     dev_props = _get_devices_properties(device_ids)
 
+    # PyTorch1.13 use multi_processor_count to check balance of devices
+    # and DataParallel of multi-devices by single-process on MLU is not supported at present.
+    # Thus we skip the imbalance check.
+    return
+
     def warn_imbalance(get_prop):
         values = [get_prop(props) for props in dev_props]
         min_pos, min_val = min(enumerate(values), key=operator.itemgetter(1))
