diff --git a/torch/csrc/utils/device_lazy_init.h b/torch/csrc/utils/device_lazy_init.h
index 9141cf3beec9..d55485bcae82 100644
--- a/torch/csrc/utils/device_lazy_init.h
+++ b/torch/csrc/utils/device_lazy_init.h
@@ -28,7 +28,7 @@ void set_requires_device_init(at::DeviceType device_type, bool value);
 
 static inline void maybe_initialize_device(at::Device& device) {
   // Add more devices here to enable lazy initialization.
-  if (device.is_cuda() || device.is_xpu()) {
+  if (device.is_cuda() || device.is_xpu() || device.is_privateuseone()) {
     device_lazy_init(device.type());
   }
 }
