diff --git a/torch/_inductor/triton_helpers.py b/torch/_inductor/triton_helpers.py
index 4f7f3145542..fb697df85b3 100644
--- a/torch/_inductor/triton_helpers.py
+++ b/torch/_inductor/triton_helpers.py
@@ -1,14 +1,27 @@
 import triton
 import triton.language as tl
 
+
 # In the latest triton, math functions were shuffled around into different modules:
 # https://github.com/openai/triton/pull/3172
-if hasattr(tl.extra.cuda, "libdevice"):
-    libdevice = tl.extra.cuda.libdevice
+try:
+    from triton.language.extra import libdevice
+
+    libdevice = tl.extra.libdevice  # noqa: F811
     math = tl.math
-else:
-    libdevice = tl.math
-    math = tl
+except ImportError:
+    if hasattr(tl.extra, "mlu") and hasattr(tl.extra.mlu, "libdevice"):
+        libdevice = tl.extra.mlu.libdevice
+        math = tl.math
+    elif hasattr(tl.extra, "cuda") and hasattr(tl.extra.cuda, "libdevice"):
+        libdevice = tl.extra.cuda.libdevice
+        math = tl.math
+    elif hasattr(tl.extra, "intel") and hasattr(tl.extra.intel, "libdevice"):
+        libdevice = tl.extra.intel.libdevice
+        math = tl.math
+    else:
+        libdevice = tl.math
+        math = tl
 
 
 @triton.jit
